src/main/java/se/liu/ida/hefquin/engine/wrappers/lpgwrapper/impl/SPARQLStar2CypherTranslatorImpl.java

```diff
@@ -806,4 +807,85 @@ protected static CypherQuery getVarVarVar(final Node s, final Node p, final Node
                         .add(new AliasedExpression(new GetItemExpression(a10, 1), gen.getRetVar(o)))
                         .build());
     }
+
+    @Override
+    public CypherMatchQuery rewriteJoins(final CypherMatchQuery query) {
+        final CypherQueryBuilder builder = new CypherQueryBuilder();
+        final List<BooleanCypherExpression> variableJoins = query.getConditions().stream()
+                .filter(x -> x instanceof EqualityExpression
+                        && ((EqualityExpression) x).getLeftExpression() instanceof CypherVar
+                        && ((EqualityExpression) x).getRightExpression() instanceof CypherVar
+                        && query.getMatchVars().contains((CypherVar) ((EqualityExpression) x).getLeftExpression())
+                        && query.getMatchVars().contains((CypherVar) ((EqualityExpression) x).getRightExpression()))
+                .collect(Collectors.toList());
+        final Map<CypherVar, CypherVar> equivalences = getEquivalenceMap(variableJoins);
+        final List<MatchClause> matches = new ArrayList<>();
+        for (final MatchClause m : query.getMatches()){
+            final MatchClause newM = (MatchClause) CypherUtils.replaceVariable(equivalences, m);
+            if (!matches.contains(newM))
+                matches.add(newM);
+        }
+        for (final BooleanCypherExpression c : query.getConditions()) {
+            if (variableJoins.contains(c)) continue;
+            builder.add(CypherUtils.replaceVariable(equivalences, c));
+        }
+        for (final UnwindIterator i : query.getIterators()) {
+            builder.add(CypherUtils.replaceVariable(equivalences, i));
+        }
+        for (final AliasedExpression r : query.getReturnExprs()) {
+            builder.add(CypherUtils.replaceVariable(equivalences, r));
+        }
+        final List<NodeMatchClause> nodes = matches.stream().filter(x->x instanceof NodeMatchClause)
+                .map(x->(NodeMatchClause)x).collect(Collectors.toList());
+        final List<MatchClause> toRemove = new ArrayList<>();
+        for (final MatchClause m : matches) {
+            if (m instanceof EdgeMatchClause) {
+                for (final NodeMatchClause node : nodes) {
+                    if (m.isRedundantWith(node))
+                        toRemove.add(node);
+                }
+            }
+        }
+        matches.removeAll(toRemove);
```

Comment:
Similar to my comment above, if the `nodes` list created in lines 838-839 is empty, there is no need to create the `toRemove` list and iterate over all the `matches`.